src/Temporalio/Activities/ActivityDefinition.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Temporalio.Activities
+{
+    /// <summary>
+    /// Definition of an activity.
+    /// </summary>
+    public class ActivityDefinition
+    {
+        private static readonly ConcurrentDictionary<Delegate, ActivityDefinition> Definitions = new();
+
+        private ActivityDefinition(string name, Delegate del)
+        {
+            Name = name;
+            Delegate = del;
+        }
+
+        /// <summary>
+        /// Gets the activity name.
+        /// </summary>
+        public string Name { get; private init; }
+
+        /// <summary>
+        /// Gets the activity delegate.
+        /// </summary>
+        public Delegate Delegate { get; private init; }
+
+        /// <summary>
+        /// Get an activity definition from a delegate containing <see cref="ActivityAttribute" />
+        /// or fail if invalid. The result is cached.
+        /// </summary>
+        /// <param name="del">Activity delegate.</param>
+        /// <returns>Activity definition.</returns>
+        /// <remarks>
+        /// Activity delegates cannot have any <c>ref</c> or <c>out</c> parameters.
+        /// </remarks>
+        public static ActivityDefinition FromDelegate(Delegate del)
+        {
+            return Definitions.GetOrAdd(del, CreateFromDelegate);
+        }
+
+        /// <summary>
+        /// Creates an activity definition from an explicit name and delegate. Most users should use
+        /// <see cref="FromDelegate" /> with attributes instead.
+        /// </summary>
+        /// <param name="name">Activity name.</param>
+        /// <param name="del">Activity delegate.</param>
+        /// <returns>Activity definition.</returns>
+        /// <seealso cref="FromDelegate" />
+        public static ActivityDefinition CreateWithoutAttribute(string name, Delegate del)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name required for activity");
```

Comment:
Consider providing the parameter name argument in these `ArgumentException` constructors. For example:
```suggestion
                throw new ArgumentException("Name required for activity", nameof(name));
```

Reply:
:+1: Will fix

Reply:
Fixed in next PR